### PR TITLE
Fixed #25933 -- Allowed an unprefixed default language in i18n_patterns().

### PR DESCRIPTION
**Pull request by urbaniak**
https://code.djangoproject.com/ticket/25933#ticket

Updated from #5823.

### CODE REVIEW

File: tests/i18n/tests.py [hunk 1]
```diff
@@ -1784,6 +1784,27 @@ def test_language_not_saved_to_session(self):
...
+        ('fr', 'French'),
+    ],
+    MIDDLEWARE_CLASSES=[
+        'django.middleware.locale.LocaleMiddleware',
+        'django.middleware.common.CommonMiddleware',
+    ],
+    ROOT_URLCONF='i18n.urls_default_unprefixed',
+    LANGUAGE_CODE='en',
+)
+class UnprefixedDefaultLanguageTests(TestCase):
```
Comment (charettes) [inline tests/i18n/tests.py]: Use `SimpleTestCase`.

File: tests/i18n/tests.py [hunk 2]
```diff
@@ -1784,6 +1784,27 @@ def test_language_not_saved_to_session(self):
...
     USE_I18N=True,
     LANGUAGES=[
+        ('en', 'English'),
+        ('fr', 'French'),
+    ],
+    MIDDLEWARE_CLASSES=[
+        'django.middleware.locale.LocaleMiddleware',
+        'django.middleware.common.CommonMiddleware',
+    ],
+    ROOT_URLCONF='i18n.urls_default_unprefixed',
```
Comment (charettes) [inline tests/i18n/tests.py]: Looks like you forgot to commit this module.

File: django/conf/urls/i18n.py [hunk 3]
```diff
@@ -4,7 +4,7 @@
 from django.views.i18n import set_language
 
 
-def i18n_patterns(*urls):
+def i18n_patterns(*urls, prefix_default_language=True):
```
Comment (charettes) [inline django/conf/urls/i18n.py]: Unfortunately this is not a valid syntax on Python 2. Use `**kwargs` and `get('prefix_default_language', True)` from it.

File: tests/i18n/urls_default_unprefixed.py [hunk 4]
```diff
@@ -0,0 +1,12 @@
+from __future__ import unicode_literals
+
+from django.conf.urls import url
+from django.conf.urls.i18n import i18n_patterns
+from django.http import HttpResponse, StreamingHttpResponse
+from django.utils.translation import ugettext_lazy as _
+
+urlpatterns = i18n_patterns(
+    url(r'^simple/$', lambda r: HttpResponse()),
```
Comment (charettes) [inline tests/i18n/urls_default_unprefixed.py]: Either add a test case for this pattern or remove it.

**Comment by urbaniak** [comment]
@charettes thanks for feedback, removed this unused pattern.

**Comment by charettes** [comment]
I wonder how this will interact with #6224.

Your `unprefixed_default_language_used` function is really similar `LocaleMiddleware.is_language_prefix_patterns_used`. Maybe we could refactor both to prevent code duplication.

 The `unprefixed_default_language_used` will also need to be `urlconf` specific.

**Comment by urbaniak** [comment]
We can have one function which will have additional param, something like that:

```python
    def is_language_prefix_patterns_used(self, urlconf, unprefixed=False):
        for url_pattern in get_resolver(urlconf).url_patterns:
            if isinstance(url_pattern, LocaleRegexURLResolver):
                if unprefixed is False or url_pattern.prefix_default_language is False:
                    return True
        return False
```

Let me know if that's right - I'll update the pull request's code.

**Comment by charettes** [comment]
That looks about right, I'll try to land #6224 soon so we can avoid merge conflicts here.

**Comment by charettes** [comment]
Ok #6224 finally got merged here's what you should be done here.

The `is_language_prefix_patterns_used` function should be declared in the `django.conf.urls.i18n` module and used by the `LocaleMiddleware` and the `get_language_from_path` function. This should prevent implementation details leaks such as the import of `LocaleRegexURLResolver`. It also should be decorated by the `lru_cache` decorator just like the `LocaleMiddleware` method is.

**Comment by urbaniak** [comment]
Rebased the branch on current master and refactored the `is_language_prefix_patterns_used` function.

File: django/utils/translation/trans_real.py [hunk 5]
```diff
@@ -465,11 +466,16 @@ def get_language_from_path(path, strict=False):
     """
     regex_match = language_code_prefix_re.match(path)
     if not regex_match:
+        if is_language_prefix_patterns_used(None, unprefixed=True):
```
Comment (charettes) [inline django/utils/translation/trans_real.py]: Hmm we can't pass `None` here, it has to be the request's `urlconf`.

I suggest we make `is_language_prefix_patterns_used` return a tuple of boolean as follow:

```python
def is_language_prefix_patterns_used(urlconf=None):
    for url_pattern in get_resolver(urlconf).url_patterns:
        if isinstance(url_pattern, LocaleRegexURLResolver):
            return True, url_pattern.prefix_default_language
        return False, False    
```

Completely remove these branch from `get_language_from_path()` and adjust `LocaleMiddleware` to correctly deal with `get_language_from_path` and `get_language_from_request` returning `None`:

```python
def process_request(self, request):
    urlconf = getattr(request, 'urlconf', settings.ROOT_URLCONF)
    i18n_patterns_used, prefixed_default_language = is_language_prefix_patterns_used(urlconf)
    language = translation.get_language_from_request(
        request, check_path=i18n_patterns_used
    )
    if not language and i18n_patterns_used and not prefixed_default_language:
        language = settings.DEFAULT_LANGUAGE
    translation.activate(language)
    request.LANGUAGE_CODE = translation.get_language()

def process_response(self, request, response):
    language = translation.get_language()
    language_from_path = translation.get_language_from_path(request.path_info)
    urlconf = getattr(request, 'urlconf', settings.ROOT_URLCONF)
    i18n_patterns_used, prefixed_default_language = is_language_prefix_patterns_used(urlconf)
    if not language_from_path and i18n_patterns_used and not prefixed_default_language:
        language_from_path = settings.DEFAULT_LANGUAGE
    if response.status_code == 404 and not language_from_path and i18n_patterns_used:
        language_path = '/%s%s' % (language, request.path_info)
        path_valid = is_valid_path(language_path, urlconf)
        path_needs_slash = (
            not path_valid and (
                settings.APPEND_SLASH and not language_path.endswith('/')
                and is_valid_path('%s/' % language_path, urlconf)
            )
        )

        if path_valid or path_needs_slash:
            script_prefix = get_script_prefix()
            # Insert language after the script prefix and before the
            # rest of the URL
            language_url = request.get_full_path(force_append_slash=path_needs_slash).replace(
                script_prefix,
                '%s%s/' % (script_prefix, language),
                1
            )
            return self.response_redirect_class(language_url)

    if not i18n_patterns_used and language_from_path:
        patch_vary_headers(response, ('Accept-Language',))
    if 'Content-Language' not in response:
        response['Content-Language'] = language
    return response
```

Do you think it makes sense?

**Comment by urbaniak** [comment]
```python
    if not i18n_patterns_used and language_from_path:
        patch_vary_headers(response, ('Accept-Language',))
```

That condition should be like the one  below.

It should set vary header only when there's no language from path, ie. / url which redirects to /en/ and /fr/.
All other responses don't vary on accept language headers cause they have got language in url.

```python
    if not (i18n_patterns_used and language_from_path):
        patch_vary_headers(response, ('Accept-Language',))
```

Except that and `s/settings.DEFAULT_LANGUAGE/settings.LANGUAGE_CODE/` it's fine. :+1:

**Comment by charettes** [comment]
These changes make sense.

File: django/middleware/locale.py [hunk 6]
```diff
@@ -21,9 +20,12 @@ class LocaleMiddleware(object):
 
     def process_request(self, request):
         urlconf = getattr(request, 'urlconf', settings.ROOT_URLCONF)
+        i18n_patterns_used, prefixed_default_language = is_language_prefix_patterns_used(urlconf)
         language = translation.get_language_from_request(
-            request, check_path=self.is_language_prefix_patterns_used(urlconf)
+            request, check_path=i18n_patterns_used
         )
+        if not language and i18n_patterns_used and not prefixed_default_language:
+            language = settings.DEFAULT_LANGUAGE
```
Comment (charettes) [inline django/middleware/locale.py]: `settings.LANGUAGE_CODE` as you pointed out.

File: django/middleware/locale.py [hunk 7]
```diff
@@ -31,8 +33,13 @@ def process_response(self, request, response):
...
         language_from_path = translation.get_language_from_path(request.path_info)
         urlconf = getattr(request, 'urlconf', settings.ROOT_URLCONF)
-        if (response.status_code == 404 and not language_from_path
-                and self.is_language_prefix_patterns_used(urlconf)):
+        i18n_patterns_used, prefixed_default_language = is_language_prefix_patterns_used(urlconf)
+        if not language_from_path and i18n_patterns_used and not prefixed_default_language:
+            language_from_path = settings.LANGUAGE_CODE
+
+        if not language_from_path and i18n_patterns_used and not prefixed_default_language:
+            language_from_path = settings.LANGUAGE_CODE
```
Comment (charettes) [inline django/middleware/locale.py]: Duplicated `if` block?

**Comment by urbaniak** [comment]
I've fixed those issues, thanks for finding them!

File: docs/releases/1.10.txt [hunk 8]
```diff
@@ -247,6 +247,7 @@ Internationalization
 * The :func:`~django.conf.urls.i18n.i18n_patterns` helper function can now be
   used in a root URLConf specified using :attr:`request.urlconf
   <django.http.HttpRequest.urlconf>`.
+* New parameter `prefix_default_language` for :func:`~django.conf.urls.i18n.i18n_patterns`.
```
Comment (charettes) [inline docs/releases/1.10.txt]: Add a new line before this item.

File: django/conf/urls/i18n.py [hunk 9]
```diff
@@ -12,7 +13,19 @@ def i18n_patterns(*urls):
...
-    return [LocaleRegexURLResolver(list(urls))]
+    prefix_default_language = kwargs.get('prefix_default_language', True)
+    return [LocaleRegexURLResolver(
+        list(urls),
+        prefix_default_language=prefix_default_language,
+    )]
+
+
+@lru_cache.lru_cache(maxsize=None)
+def is_language_prefix_patterns_used(urlconf):
```
Comment (charettes) [inline django/conf/urls/i18n.py]: A small docstring wouldn't hurt here.

**Comment by charettes** [comment]
The patch is looking good, I'll mark it ready for a final review.

File: docs/topics/i18n/translation.txt [hunk 10]
```diff
@@ -1371,6 +1377,17 @@ function. Example::
     >>> reverse('news:detail', kwargs={'slug': 'news-slug'})
     '/nl/news/news-slug/'
 
+With ``prefix_default_language`` set to ``False`` the urls will be::
```
Comment (timgraham) [inline docs/topics/i18n/translation.txt]: URLs

File: docs/topics/i18n/translation.txt [hunk 11]
```diff
@@ -1371,6 +1377,17 @@ function. Example::
...
+
+    >>> activate('en')
+    >>> reverse('news:index')
+    '/news/'
+
+    >>> activate('nl')
+    >>> reverse('news:index')
+    '/nl/news/'
+
```
Comment (timgraham) [inline docs/topics/i18n/translation.txt]: Need a block here:
```
.. versionadded:: 1.10

    The ``prefix_default_language`` parameter was added.
```

File: docs/releases/1.10.txt [hunk 12]
```diff
@@ -248,6 +248,8 @@ Internationalization
   used in a root URLConf specified using :attr:`request.urlconf
   <django.http.HttpRequest.urlconf>`.
 
+* New parameter `prefix_default_language` for :func:`~django.conf.urls.i18n.i18n_patterns`.
```
Comment (timgraham) [inline docs/releases/1.10.txt]: It's useful to describe what the parameter does, e.g. "The new ``prefix_default_language`` parameter for :func:`~django.conf.urls.i18n.i18n_patterns` allows..." (also wrap lines at 79 characters)

File: django/urls/resolvers.py [hunk 13]
```diff
@@ -381,15 +382,23 @@ class LocaleRegexURLResolver(RegexURLResolver):
     Rather than taking a regex argument, we just override the ``regex``
     function to always return the active language-code as regex.
     """
-    def __init__(self, urlconf_name, default_kwargs=None, app_name=None, namespace=None):
+    def __init__(
+        self, urlconf_name, default_kwargs=None, app_name=None, namespace=None,
+        prefix_default_language=True
```
Comment (timgraham) [inline django/urls/resolvers.py]: include a trailing comma

File: django/conf/urls/i18n.py [hunk 14]
```diff
@@ -12,7 +13,24 @@ def i18n_patterns(*urls):
...
+    """
+    Returns tuple of two booleans, first is `True` when
```
Comment (timgraham) [inline django/conf/urls/i18n.py]: Return a tuple...

File: docs/topics/i18n/translation.txt [hunk 15]
```diff
@@ -1326,11 +1326,17 @@ Django provides two mechanisms to internationalize URL patterns:
...
-.. function:: i18n_patterns(*pattern_list)
+.. function:: i18n_patterns(*pattern_list, prefix_default_language=True)
 
 This function can be used in a root URLconf and Django will automatically
 prepend the current active language code to all url patterns defined within
-:func:`~django.conf.urls.i18n.i18n_patterns`. Example URL patterns::
+:func:`~django.conf.urls.i18n.i18n_patterns`.
+
+Setting ``prefix_default_language`` to ``False`` will remove prefix from the
+default language. This can be useful when adding translations to existing site,
```
Comment (timgraham) [inline docs/topics/i18n/translation.txt]:
```
default language (:setting:`LANGUAGE_CODE`).
```

File: docs/topics/i18n/translation.txt [hunk 16]
```diff
@@ -1326,11 +1326,17 @@ Django provides two mechanisms to internationalize URL patterns:
...
+so the current urls won't change.
```
Comment (timgraham) [inline docs/topics/i18n/translation.txt]: URLs

File: docs/topics/i18n/translation.txt [hunk 17]
```diff
@@ -1326,11 +1326,17 @@ Django provides two mechanisms to internationalize URL patterns:
...
 
-.. function:: i18n_patterns(*pattern_list)
+.. function:: i18n_patterns(*pattern_list, prefix_default_language=True)
 
 This function can be used in a root URLconf and Django will automatically
 prepend the current active language code to all url patterns defined within
-:func:`~django.conf.urls.i18n.i18n_patterns`. Example URL patterns::
+:func:`~django.conf.urls.i18n.i18n_patterns`.
+
+Setting ``prefix_default_language`` to ``False`` will remove prefix from the
```
Comment (timgraham) [inline docs/topics/i18n/translation.txt]: will remove -> removes the

**Comment by timgraham** [comment]
Please squash commits and use the commit message format from the pull request title.

**Comment by urbaniak** [comment]
Squashed those commits. Is it all right?

**Comment by timgraham** [comment]
Sent a PR with a few cosmetic edits: https://github.com/urbaniak/django/pull/1 (feel free to squash it in) otherwise looks good!

**Comment by urbaniak** [comment]
Merged that and squashed, thanks!

**Comment by timgraham** [comment]
merged in 839a955d08fa0657622ed9b60fdd262658fa659f, thanks!

**Pull request closed without merging**